src/index.js

```
@@ -1,41 +1,194 @@
+    })
+  }
+
+  this.stop = (callback) => {
+    this.isOnline = false
+    this.swarm.close(callback)
+  }
+
+  this.dialById = (id, protocol, callback) => {
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+
+    if (!this.isOnline) {
+      return callback(OFFLINE_ERROR)
+    }
+    // NOTE, these dialById only works if a previous dial
+    // was made until we have PeerRouting
+    // TODO support PeerRouting when it is Ready
+    callback(new Error('not implemented yet'))
+  }
+
+  this.dialByMultiaddr = (maddr, protocol, callback) => {
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+
+    if (!this.isOnline) {
+      return callback(OFFLINE_ERROR)
+    }
+
+    if (typeof maddr === 'string') {
+      maddr = multiaddr(maddr)
+    }
+
+    if (!mafmt.IPFS.matches(maddr.toString())) {
+      return callback(new Error('multiaddr not valid'))
+    }
+
+    let ipfsIdB58String
+    maddr.stringTuples().forEach((tuple) => {
+      if (tuple[0] === 421) {
+        ipfsIdB58String = tuple[1]
+      }
+    })
+
+    try {
+      const pi = this.peerBook.getByB58String(ipfsIdB58String)
+      pi.multiaddr.add(maddr)
+      this.dialByPeerInfo(pi, protocol, callback)
+    } catch (err) {
+      const pId = PeerId.createFromB58String(ipfsIdB58String)
+      const peer = new PeerInfo(pId)
+      peer.multiaddr.add(maddr)
+      this.dialByPeerInfo(peer, protocol, callback)
+    }
+  }
+
+  this.dialByPeerInfo = (peer, protocol, callback) => {
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+    if (!this.isOnline) {
+      return callback(OFFLINE_ERROR)
+    }
+
+    this.swarm.dial(peer, protocol, (err, conn) => {
+      if (err) {
+        return callback(err)
+      }
+      this.peerBook.put(peer)
+      callback(null, conn)
+    })
+  }
+
+  this.hangUpById = (id, callback) => {
+    // TODO
```

maybe return an error for now?